include/oneapi/mkl/rng/device/detail/gaussian_impl.hpp

```diff
@@ -0,0 +1,270 @@
+            if (!flag_) {
+                constexpr std::int32_t tail = EngineType::vec_size % 2;
+                auto uniform_res = engine.generate(RealType(0), RealType(1));
+#pragma unroll
+                for (std::int32_t i = 0; i < EngineType::vec_size - tail; i += 2) {
+                    u1 = uniform_res[i];
+                    u2 = uniform_res[i + 1];
+                    u1_transformed = ln_wrapper(u1);
+                    u1_transformed = sqrt_wrapper(static_cast<RealType>(-2.0) * u1_transformed);
+                    sin = sincospi_wrapper(RealType(2.0) * u2, cos);
+                    res[i] = (u1_transformed * sin) * stddev_ + mean_;
+                    res[i + 1] = (u1_transformed * cos) * stddev_ + mean_;
+                }
+                if constexpr (tail) {
+                    u1 = uniform_res[EngineType::vec_size - 1];
+                    u2 = engine.generate_single(RealType(0), RealType(1));
+                    u1_transformed = ln_wrapper(u1);
+                    u1_transformed = sqrt_wrapper(static_cast<RealType>(-2.0) * u1_transformed);
+                    res[EngineType::vec_size - 1] =
+                        u1_transformed * sinpi_wrapper(RealType(2) * u2) * stddev_ + mean_;
+                    u1_transformed_ = u1_transformed;
+                    u2_ = u2;
+                    flag_ = true;
+                }
+                return res;
+            }
+
+            res[0] = u1_transformed_ * cospi_wrapper(RealType(2) * u2_) * stddev_ + mean_;
+            flag_ = false;
+            constexpr std::int32_t tail = (EngineType::vec_size - 1) % 2;
+#pragma unroll
+            for (std::int32_t i = 1; i < EngineType::vec_size - tail; i += 2) {
+                u1 = engine.generate_single(RealType(0), RealType(1));
+                u2 = engine.generate_single(RealType(0), RealType(1));
+                u1_transformed = ln_wrapper(u1);
+                u1_transformed = sqrt_wrapper(static_cast<RealType>(-2.0) * u1_transformed);
+                sin = sincospi_wrapper(RealType(2.0) * u2, cos);
+                res[i] = (u1_transformed * sin) * stddev_ + mean_;
+                res[i + 1] = (u1_transformed * cos) * stddev_ + mean_;
+            }
+            if constexpr (tail) {
+                u1 = engine.generate_single(RealType(0), RealType(1));
+                u2 = engine.generate_single(RealType(0), RealType(1));
+                u1_transformed = ln_wrapper(u1);
+                u1_transformed = sqrt_wrapper(static_cast<RealType>(-2.0) * u1_transformed);
+                res[EngineType::vec_size - 1] =
+                    u1_transformed * sinpi_wrapper(RealType(2) * u2) * stddev_ + mean_;
+                u1_transformed_ = u1_transformed;
+                u2_ = u2;
+                flag_ = true;
+            }
+            return res;
+        }
+    }
+
+    template <typename EngineType>
+    __attribute__((always_inline)) inline RealType generate_single(EngineType& engine) {
```

Comment:
we can remove generate_single here as it is not part of the spec (but engine_impl should keep it for lines 147-148 in gaussian impl)

Reply:
We've decided to keep internal generate_single methods